### PR TITLE
abci: implement process proposal to spec

### PR DESCRIPTION
**Pull request by cmwaters**
Closes: #9121

Ref: #9053

### CODE REVIEW

**Comment by cmwaters** [comment]
Looking through the changes, I think we're light on tests in the consensus package. I would like to mock out the application and test some variations of the new calls but I want to do this in a separate PR.

File: state/execution.go [hunk 1]
```diff
@@ -334,7 +339,7 @@ func execBlockOnProxyApp(
 	}
 	proxyAppConn.SetResponseCallback(proxyCb)
 
-	commitInfo := getBeginBlockValidatorInfo(block, store, initialHeight)
+	commitInfo := buildLastCommitInfo(block, store, initialHeight)
 
 	byzVals := make([]abci.Evidence, 0)
```
Comment (sergio-mena) [inline state/execution.go:344]: The original PR was removing lines 344-347. I don't really remember why, could you please check whether these should be removed?

File: state/execution_test.go [hunk 2]
```diff
@@ -233,7 +233,7 @@ func TestProcessProposal(t *testing.T) {
```
Comment (sergio-mena) [inline state/execution_test.go:233]: The original PR introduced some improvements to test case `TestProcessProposal`.
Consider bringing these improvements along (a few 10s of lines)

File: consensus/mempool_test.go [hunk 3]
```diff
@@ -261,3 +261,8 @@ func (app *CounterApplication) PrepareProposal(
 	req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
```
Comment (sergio-mena) [inline consensus/mempool_test.go:261]: There are two files that contain changes in the original PR:
* (internal)/consensus/common_test.go These changes change `makeState` but it's not clear to me how important that is (it seems a convenience change). @williambanfield, what to do think?
* (internal)/consensus/state_test.go In this file, test case `TestProcessProposalAccept` is written from scratch. It seems relevant for the code of this PR. Please port it to this PR

**Review by sergio-mena — COMMENTED** [review]
Awesome job!

File: consensus/mempool_test.go (hunk 3 above)
Comment (williambanfield) [reply consensus/mempool_test.go:261]: :+1: to the second change. As per the first change, that actually is necessary to make the first change doable, so both should be done.

**Review by williambanfield — CHANGES_REQUESTED** [review]
Should port the changes @sergio-mena suggested. Looks great beyond that.

File: consensus/mempool_test.go (hunk 3 above)
Comment (cmwaters) [reply consensus/mempool_test.go]: Lol I didn't realise how much clean up work was done in these tests between v0.34 and v0.35

Comment (cmwaters) [reply consensus/mempool_test.go]: This is ridiculous

File: rpc/jsonrpc/server/http_server.go [hunk 4]
```diff
@@ -53,10 +53,11 @@ func DefaultConfig() *Config {
 func Serve(listener net.Listener, handler http.Handler, logger log.Logger, config *Config) error {
 	logger.Info("serve", "msg", log.NewLazySprintf("Starting RPC HTTP server on %s", listener.Addr()))
 	s := &http.Server{
-		Handler:        RecoverAndLogHandler(maxBytesHandler{h: handler, n: config.MaxBodyBytes}, logger),
-		ReadTimeout:    config.ReadTimeout,
-		WriteTimeout:   config.WriteTimeout,
-		MaxHeaderBytes: config.MaxHeaderBytes,
+		Handler:           RecoverAndLogHandler(maxBytesHandler{h: handler, n: config.MaxBodyBytes}, logger),
+		ReadTimeout:       config.ReadTimeout,
+		ReadHeaderTimeout: config.ReadTimeout,
```
Comment (williambanfield) [inline rpc/jsonrpc/server/http_server.go:58]: How is this related? I think we should keep changes scoped just to the ABCI++ work for now.

Comment (williambanfield) [reply rpc/jsonrpc/server/http_server.go:58]: It also looks like, from the Go docs, this is not necessary, since ReadHeaderTimeout takes the value of ReadTimeout if not set. 

https://pkg.go.dev/net/http#Server

File: p2p/transport_test.go [hunk 5]
```diff
@@ -517,9 +517,9 @@ func TestTransportMultiplexRejectSelf(t *testing.T) {
 	}()
 
 	if err := <-errc; err != nil {
-		if err, ok := err.(ErrRejected); ok {
-			if !err.IsSelf() {
-				t.Errorf("expected to reject self, got: %v", err)
+		if e, ok := err.(ErrRejected); ok {
```
Comment (williambanfield) [inline p2p/transport_test.go:520]: Not sure why this change is made here, but think we should keep the scope of this PR just to ABCI++.

File: test/maverick/node/node.go [hunk 6]
```diff
@@ -1215,6 +1215,7 @@ func (n *Node) startPrometheusServer(addr string) *http.Server {
 				promhttp.HandlerOpts{MaxRequestsInFlight: n.config.Instrumentation.MaxOpenConnections},
 			),
 		),
+		ReadHeaderTimeout: 10 * time.Second,
```
Comment (williambanfield) [inline test/maverick/node/node.go:1218]: Similarly unrelated change.

File: node/node.go [hunk 7]
```diff
@@ -69,6 +69,8 @@ type DBContext struct {
 // DBProvider takes a DBContext and returns an instantiated DB.
 type DBProvider func(*DBContext) (dbm.DB, error)
 
+const readHeaderTimeout = 10 * time.Second
```
Comment (williambanfield) [inline node/node.go:72]: Not sure is related, same with other change in this file.

**Review by williambanfield — CHANGES_REQUESTED** [review]
It looks like a few changes unrelated to ABCI++ were mixed in here, some of which have already landed in `main` . I don't think they are necessary to do again here since this will be merged into `main` anyway.

A few notes otherwise, but mostly looks good.

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -22,7 +22,7 @@ jobs:
     steps:
       - uses: actions/setup-go@v3
         with:
-          go-version: "1.17"
+          go-version: "1.18"
```
[the remark this answers is not in the document]
Comment (cmwaters) [reply .github/workflows/build.yml:25]: I just cherry-picked a commit from main into here which bumped the go version and did a bunch of linting changes to appease the linter

File: abci/types/types.go [hunk 9]
```diff
@@ -41,6 +41,11 @@ func (r ResponseQuery) IsErr() bool {
 	return r.Code != CodeTypeOK
 }
 
+// IsOK returns true if Code is OK
+func (r ResponseProcessProposal) IsOK() bool {
```
[the remark this answers is not in the document]
Comment (cmwaters) [reply abci/types/types.go:45]: Sure. I don't mind doing that

File: state/execution.go (hunk 1 above)
Comment (cmwaters) [reply state/execution.go:355]: Yeah I saw that. I wasn't planning to add that commit until after. We also changed abci.Evidence to abci.Misbehavior

File: state/execution.go [hunk 10]
```diff
@@ -149,6 +149,27 @@ func (blockExec *BlockExecutor) CreateProposalBlock(
...
+	block *types.Block,
+	state State,
+) (bool, error) {
+	resp, err := blockExec.proxyApp.ProcessProposalSync(abci.RequestProcessProposal{
+		Hash:               block.Header.Hash(),
+		Height:             block.Header.Height,
+		Time:               block.Header.Time,
+		Txs:                block.Data.Txs.ToSliceOfBytes(),
+		ProposedLastCommit: buildLastCommitInfo(block, blockExec.store, state.InitialHeight),
+		Misbehavior:        block.Evidence.Evidence.ToABCI(),
```
[the remark this answers is not in the document]
Comment (cmwaters) [reply state/execution.go:162]: I was going to do that in a separate commit. Sergio and I discussed this offline

File: rpc/jsonrpc/server/http_server.go (hunk 4 above)
Comment (cmwaters) [reply rpc/jsonrpc/server/http_server.go:58]: Yeah it was picked up by the l'Inter I believe

Comment (williambanfield) [reply rpc/jsonrpc/server/http_server.go:58]: Hmm, mind pointing me toward the lint error? Not sure why we have a linter recommending we set this.

**Review by williambanfield — APPROVED** [review]
